// @blocklet/sdk/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.config = exports.Security = exports.Component = exports.component = exports.middlewares = exports.env = exports.getWallet = exports.Database = exports.connectShared = exports.createConnectHandlers = exports.BlockletAuthenticator = exports.WalletAuthenticator = exports.WalletHandlers = exports.Notification = exports.NotificationService = exports.Auth = exports.AuthService = void 0;
const auth_1 = __importDefault(require("./service/auth"));
exports.AuthService = auth_1.default;
exports.Auth = auth_1.default;
const notification_1 = __importDefault(require("./service/notification"));
exports.NotificationService = notification_1.default;
exports.Notification = notification_1.default;
const wallet_authenticator_1 = __importDefault(require("./wallet-authenticator"));
exports.WalletAuthenticator = wallet_authenticator_1.default;
const wallet_handler_1 = __importDefault(require("./wallet-handler"));
exports.WalletHandlers = wallet_handler_1.default;
const authenticator_1 = __importDefault(require("./connect/authenticator"));
exports.BlockletAuthenticator = authenticator_1.default;
const handler_1 = __importDefault(require("./connect/handler"));
exports.createConnectHandlers = handler_1.default;
const shared_1 = __importDefault(require("./connect/shared"));
exports.connectShared = shared_1.default;
const index_1 = __importDefault(require("./database/index"));
exports.Database = index_1.default;
const env_1 = __importDefault(require("./env"));
exports.env = env_1.default;
const index_2 = __importDefault(require("./middlewares/index"));
exports.middlewares = index_2.default;
const wallet_1 = __importDefault(require("./wallet"));
exports.getWallet = wallet_1.default;
const index_3 = __importDefault(require("./component/index"));
exports.component = index_3.default;
exports.Component = index_3.default;
const index_4 = __importDefault(require("./security/index"));
exports.Security = index_4.default;
const config_1 = __importDefault(require("./config"));
exports.config = config_1.default;
__exportStar(require("./component"), exports);
