// @ocap/mcrypto/lib/crypter/rsa.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RSACrypter = void 0;
const crypto_1 = __importDefault(require("crypto"));
const util_1 = require("@ocap/util");
const crypter_1 = __importDefault(require("../protocols/crypter"));
const encode_1 = require("../encode");
// RSA-OAEP
class RSACrypter extends crypter_1.default {
    genKeyPair(length = 2048) {
        return crypto_1.default.generateKeyPairSync('rsa', {
            modulusLength: length,
            publicKeyEncoding: {
                type: 'spki',
                format: 'pem',
            },
            privateKeyEncoding: {
                type: 'pkcs8',
                format: 'pem',
            },
        });
    }
    encrypt(message, key, encoding = 'hex') {
        return (0, encode_1.encode)(crypto_1.default.publicEncrypt(key, (0, util_1.toBuffer)(message)), encoding);
    }
    decrypt(message, key, encoding = 'hex') {
        return (0, encode_1.encode)(crypto_1.default.privateDecrypt(key, (0, util_1.toBuffer)(message)), encoding);
    }
}
exports.RSACrypter = RSACrypter;
exports.default = new RSACrypter();
