// @nedb/multi/lib/cursor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Cursor = void 0;
const rpc_1 = require("./rpc");
const constants_1 = require("./constants");
class Cursor {
    constructor(socket, options, args) {
        this.options = options;
        this.socket = socket;
        this._query = args[0] || {};
    }
    query(query) {
        this._query = query;
        return this;
    }
    sort(sort) {
        this._sort = sort;
        return this;
    }
    skip(skip) {
        this._skip = skip;
        return this;
    }
    limit(limit) {
        this._limit = limit;
        return this;
    }
    projection(projection) {
        this._projection = projection;
        return this;
    }
    toJSON() {
        return {
            query: this._query,
            skip: this._skip,
            sort: this._sort,
            limit: this._limit,
            projection: this._projection,
        };
    }
    exec(callback) {
        return (0, rpc_1.doRpc)(this.socket, this.options, constants_1.EXECUTE_CURSOR_PRIVATE, [this, callback].filter(Boolean));
    }
}
exports.Cursor = Cursor;
