// @blocklet/meta/lib/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.verifyMultiSig = exports.getBlockletEngine = exports.getBlockletInfo = exports.getBlockletWallet = exports.toBlockletDid = exports.fixService = exports.fixInterfaces = exports.fixPerson = exports.parsePerson = exports.formatPerson = exports.fixAndValidateService = exports.validateMeta = exports.parse = exports.update = exports.select = exports.list = exports.constants = void 0;
const constants_1 = __importDefault(require("./constants"));
exports.constants = constants_1.default;
const parse_1 = __importDefault(require("./parse"));
exports.parse = parse_1.default;
const did_1 = __importDefault(require("./did"));
exports.toBlockletDid = did_1.default;
const wallet_1 = __importDefault(require("./wallet"));
exports.getBlockletWallet = wallet_1.default;
const info_1 = __importDefault(require("./info"));
exports.getBlockletInfo = info_1.default;
const engine_1 = __importDefault(require("./engine"));
exports.getBlockletEngine = engine_1.default;
const validate_1 = require("./validate");
Object.defineProperty(exports, "validateMeta", { enumerable: true, get: function () { return validate_1.validateMeta; } });
Object.defineProperty(exports, "fixAndValidateService", { enumerable: true, get: function () { return validate_1.fixAndValidateService; } });
const fix_1 = require("./fix");
Object.defineProperty(exports, "formatPerson", { enumerable: true, get: function () { return fix_1.formatPerson; } });
Object.defineProperty(exports, "parsePerson", { enumerable: true, get: function () { return fix_1.parsePerson; } });
Object.defineProperty(exports, "fixPerson", { enumerable: true, get: function () { return fix_1.fixPerson; } });
Object.defineProperty(exports, "fixInterfaces", { enumerable: true, get: function () { return fix_1.fixInterfaces; } });
Object.defineProperty(exports, "fixService", { enumerable: true, get: function () { return fix_1.fixService; } });
const file_1 = require("./file");
Object.defineProperty(exports, "list", { enumerable: true, get: function () { return file_1.list; } });
Object.defineProperty(exports, "select", { enumerable: true, get: function () { return file_1.select; } });
Object.defineProperty(exports, "update", { enumerable: true, get: function () { return file_1.update; } });
const verify_multi_sig_1 = __importDefault(require("./verify-multi-sig"));
exports.verifyMultiSig = verify_multi_sig_1.default;
exports.default = {
    constants: constants_1.default,
    list: file_1.list,
    select: file_1.select,
    update: file_1.update,
    parse: parse_1.default,
    validateMeta: validate_1.validateMeta,
    fixAndValidateService: validate_1.fixAndValidateService,
    formatPerson: fix_1.formatPerson,
    parsePerson: fix_1.parsePerson,
    fixPerson: fix_1.fixPerson,
    fixInterfaces: fix_1.fixInterfaces,
    fixService: fix_1.fixService,
    toBlockletDid: did_1.default,
    getBlockletWallet: wallet_1.default,
    getBlockletInfo: info_1.default,
    getBlockletEngine: engine_1.default,
    verifyMultiSig: verify_multi_sig_1.default,
};
