// @blocklet/sdk/lib/util/env.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseEnv = void 0;
const constant_1 = require("@blocklet/constant");
function parseEnv(env = process.env) {
    const preferences = Object.keys(env)
        .filter((x) => x.startsWith(constant_1.BLOCKLET_PREFERENCE_PREFIX))
        .reduce((acc, x) => {
        const key = x.replace(constant_1.BLOCKLET_PREFERENCE_PREFIX, '');
        try {
            acc[key] = JSON.parse(env[x]);
        }
        catch (_a) {
            acc[key] = env[x];
        }
        return acc;
    }, {});
    return Object.freeze({
        appId: env.BLOCKLET_APP_ID,
        appName: env.BLOCKLET_APP_NAME,
        appDescription: env.BLOCKLET_APP_DESCRIPTION,
        appUrl: env.BLOCKLET_APP_URL,
        isComponent: env.BLOCKLET_DID !== env.BLOCKLET_REAL_DID,
        dataDir: env.BLOCKLET_DATA_DIR,
        cacheDir: env.BLOCKLET_CACHE_DIR,
        mode: env.BLOCKLET_MODE,
        appStorageEndpoint: env.BLOCKLET_APP_SPACE_ENDPOINT,
        preferences,
    });
}
exports.parseEnv = parseEnv;
